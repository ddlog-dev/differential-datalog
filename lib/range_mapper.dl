extern type Mapper<'K, 'V>
extern type Mapping<'K, 'V>

extern function new_mapper(name: string, start: 'V, end: 'V): Mapper<'K, 'V>

extern function get_mapping(mapper: Mapper<'K, 'V>, k: 'K): Mapping<'K, 'V>

extern function mkey(m: Mapping<'K, 'V>): 'K
extern function mval(m: Mapping<'K, 'V>): Option<'V>
extern function mmapper_name(m: Mapping<'K, 'V>): string
