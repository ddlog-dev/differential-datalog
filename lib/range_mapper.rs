use range_interner as ri;

#[derive(Eq, PartialEq, Clone)]
pub struct range_mapper_Mapper<K: ri::RangeInternerKey, V: ri::RangeInternerVal<V>> {
    interner: ri::RangeInterner<K, V>
}

impl<K: ri::RangeInternerKey + std::fmt::Debug, V: ri::RangeInternerVal<V> + std::fmt::Debug> std::fmt::Debug for range_mapper_Mapper<K, V> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        std::fmt::Debug::fmt(&self.interner, f)
    }
}

impl<K: ri::RangeInternerKey + std::fmt::Debug, V: ri::RangeInternerVal<V> + std::fmt::Debug> std::fmt::Display for range_mapper_Mapper<K, V> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        std::fmt::Debug::fmt(&self.interner, f)
    }
}

impl<K: ri::RangeInternerKey, V: ri::RangeInternerVal<V>> Default for range_mapper_Mapper<K, V> {
    fn default() -> Self {
        range_mapper_Mapper {
            interner: ri::RangeInterner::new("default", std::ops::Range{start: V::one(), end: V::one()})
        }
    }
}

impl<K: ri::RangeInternerKey, V: ri::RangeInternerVal<V>> serde::Serialize for range_mapper_Mapper<K, V> {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: serde::Serializer,
    {
        serializer.serialize_str(self.interner.name().as_str())
    }
}

impl<'de, K: ri::RangeInternerKey, V: ri::RangeInternerVal<V>> serde::Deserialize<'de> for range_mapper_Mapper<K, V> {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: serde::Deserializer<'de>,
    {
        Err(<D::Error as serde::de::Error>::custom("range_mapper.Mapper.deserialize() is not implemented"))
    }
}

impl<K: ri::RangeInternerKey, V: ri::RangeInternerVal<V>> FromRecord for range_mapper_Mapper<K, V> {
    fn from_record(val: &record::Record) -> Result<Self, String> {
        Err("range_mapper.Mapper.from_record() is not implemented".to_string())
    }
}

impl<K: ri::RangeInternerKey, V: ri::RangeInternerVal<V>> IntoRecord for range_mapper_Mapper<K, V> {
    fn into_record(self) -> record::Record {
        record::Record::String(self.interner.name())
    }
}

impl<K: ri::RangeInternerKey, V: ri::RangeInternerVal<V>> record::Mutator<range_mapper_Mapper<K,V>> for record::Record {
    fn mutate(&self, vec: &mut range_mapper_Mapper<K, V>) -> Result<(), String> {
        Err("range_mapper.Mapper.mutate() is not implemented".to_string())
    }
}

impl<K: ri::RangeInternerKey, V: ri::RangeInternerVal<V>> PartialOrd for range_mapper_Mapper<K, V>
{
    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
        let sptr = self.interner.inner().as_ref() as *const std::sync::Mutex<_> as usize;
        let optr = other.interner.inner().as_ref() as *const std::sync::Mutex<_> as usize;
        sptr.partial_cmp(&optr)
    }
}

impl<K: ri::RangeInternerKey, V: ri::RangeInternerVal<V>> Ord for range_mapper_Mapper<K, V>
{
    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
        let sptr = self.interner.inner().as_ref() as *const std::sync::Mutex<_> as usize;
        let optr = other.interner.inner().as_ref() as *const std::sync::Mutex<_> as usize;
        sptr.cmp(&optr)
    }
}

impl<K: ri::RangeInternerKey, V: ri::RangeInternerVal<V>> std::hash::Hash for range_mapper_Mapper<K, V> {
    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
        (self.interner.inner().as_ref() as *const std::sync::Mutex<_>).hash(state);
    }
}

#[derive(Eq, PartialOrd, PartialEq, Ord, Clone, Hash)]
pub struct range_mapper_Mapping<K: ri::RangeInternerKey, V: ri::RangeInternerVal<V>> {
    intern: ri::RangeIntern<K, V>
}

impl<K: ri::RangeInternerKey + std::fmt::Debug, V: ri::RangeInternerVal<V> + std::fmt::Debug> std::fmt::Debug for range_mapper_Mapping<K, V> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        std::fmt::Debug::fmt(&self.intern, f)
    }
}

impl<K: ri::RangeInternerKey + std::fmt::Display + Clone, V: ri::RangeInternerVal<V> + std::fmt::Display> std::fmt::Display for range_mapper_Mapping<K, V> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        f.write_fmt(format_args!("({}->{})", range_mapper_mkey(self), range_mapper_mval(self)))
    }
}

impl<K: ri::RangeInternerKey + Default + Clone, V: ri::RangeInternerVal<V>> Default for range_mapper_Mapping<K, V> {
    fn default() -> Self {
        range_mapper_get_mapping(&range_mapper_Mapper::default(), &K::default())
    }
}

impl<K: ri::RangeInternerKey, V: ri::RangeInternerVal<V>> serde::Serialize for range_mapper_Mapping<K, V> {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: serde::Serializer,
    {
        Err(<S::Error as serde::ser::Error>::custom("range_mapper.Mapping.serialize() is not implemented"))
    }
}

impl<'de, K: ri::RangeInternerKey, V: ri::RangeInternerVal<V>> serde::Deserialize<'de> for range_mapper_Mapping<K, V> {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: serde::Deserializer<'de>,
    {
        Err(<D::Error as serde::de::Error>::custom("range_mapper.Mapping.deserialize() is not implemented"))
    }
}

impl<K: ri::RangeInternerKey, V: ri::RangeInternerVal<V>> FromRecord for range_mapper_Mapping<K, V> {
    fn from_record(val: &record::Record) -> Result<Self, String> {
        Err("range_mapper.Mapping.from_record() is not implemented".to_string())
    }
}

impl<K: ri::RangeInternerKey + IntoRecord + Clone, V: ri::RangeInternerVal<V> + IntoRecord> IntoRecord for range_mapper_Mapping<K, V> {
    fn into_record(self) -> record::Record {
        record::Record::Tuple(vec![range_mapper_mkey(&self).into_record(), range_mapper_mval(&self).into_record()])
    }
}

impl<K: ri::RangeInternerKey, V: ri::RangeInternerVal<V>> record::Mutator<range_mapper_Mapping<K,V>> for record::Record {
    fn mutate(&self, vec: &mut range_mapper_Mapping<K, V>) -> Result<(), String> {
        Err("range_mapper.Mapping.mutate() is not implemented".to_string())
    }
}

pub fn range_mapper_new_mapper<K, V>(name: &String, start: &V, end: &V) -> range_mapper_Mapper<K, V>
where
K: ri::RangeInternerKey,
V: ri::RangeInternerVal<V>
{
    range_mapper_Mapper{
        interner: ri::RangeInterner::new(name.as_str(), std::ops::Range{start: *start, end: *end})
    }
}

pub fn range_mapper_get_mapping<K, V>(mapper: &range_mapper_Mapper<K, V>, k: &K) -> range_mapper_Mapping<K, V>
where
K: ri::RangeInternerKey + Clone,
V: ri::RangeInternerVal<V>
{
    range_mapper_Mapping {
        intern: mapper.interner.intern((*k).clone())
    }
}

pub fn range_mapper_mkey<K, V>(m: &range_mapper_Mapping<K, V>) -> K
where
K: ri::RangeInternerKey + Clone,
V: ri::RangeInternerVal<V>
{
    m.intern.key().clone()
}

pub fn range_mapper_mval<K, V>(m: &range_mapper_Mapping<K, V>) -> std_Option<V>
where
K: ri::RangeInternerKey,
V: ri::RangeInternerVal<V>
{
    option2std(m.intern.val())
}

pub fn range_mapper_mmapper_name<K, V>(m: &range_mapper_Mapping<K, V>) -> String
where
K: ri::RangeInternerKey,
V: ri::RangeInternerVal<V>
{
    m.intern.interner().name()
}
