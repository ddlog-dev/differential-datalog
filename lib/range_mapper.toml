[dependencies.range-interner]
path = "/home/lryzhyk/projects/range-interner"
