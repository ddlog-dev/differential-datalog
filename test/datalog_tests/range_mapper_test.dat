dump range_mapper_test.RangeMappingKV;

# One new mapping per transaction to keep the order deterministic.

start;
# already has a static mapping
insert range_mapper_test.Request(100),
# new mapping
insert range_mapper_test.Request(1000),
commit dump_changes;

start;
# new mapping
insert range_mapper_test.Request(2000),
commit dump_changes;

start;
# new mapping
insert range_mapper_test.Request(3000),
commit dump_changes;

start;
# new mapping
insert range_mapper_test.Request(4000),
commit dump_changes;

start;
# new mapping
insert range_mapper_test.Request(5000),
commit dump_changes;

start;
# we should run out of range here
insert range_mapper_test.Request(6000),
commit dump_changes;
