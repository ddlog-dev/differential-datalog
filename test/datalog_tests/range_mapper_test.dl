import range_mapper

relation RangeMapper(mapper: Mapper<u128, u16>)

RangeMapper(new_mapper("test_mapper", 16'd0, 16'd10)).

relation RangeMapping(mapping: Mapping<u128, u16>)

RangeMapping(m1),
RangeMapping(m2),
RangeMapping(m3),
RangeMapping(m4),
RangeMapping(m5) :-
    RangeMapper(mapper),
    var m1 = get_mapping(mapper, 128'd100),
    var m2 = get_mapping(mapper, 128'd200),
    var m3 = get_mapping(mapper, 128'd300),
    var m4 = get_mapping(mapper, 128'd400),
    var m5 = get_mapping(mapper, 128'd500).

input relation Request(r: u128)

RangeMapping(m) :-
    RangeMapper(mapper),
    Request(r),
    var m = get_mapping(mapper, r).

output relation RangeMappingKV(mapper: string,  k: u128, v: Option<u16>)

RangeMappingKV(mmapper_name(m), mkey(m), mval(m)) :-
    RangeMapping(m),
    RangeMapping(m).
