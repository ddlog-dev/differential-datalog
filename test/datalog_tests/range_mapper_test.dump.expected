range_mapper_test.RangeMappingKV{.mapper = "test_mapper", .k = 100, .v = std.Some{.x = 0}}
range_mapper_test.RangeMappingKV{.mapper = "test_mapper", .k = 200, .v = std.Some{.x = 1}}
range_mapper_test.RangeMappingKV{.mapper = "test_mapper", .k = 300, .v = std.Some{.x = 2}}
range_mapper_test.RangeMappingKV{.mapper = "test_mapper", .k = 400, .v = std.Some{.x = 3}}
range_mapper_test.RangeMappingKV{.mapper = "test_mapper", .k = 500, .v = std.Some{.x = 4}}
range_mapper_test.RangeMappingKV:
range_mapper_test.RangeMappingKV{.mapper = "test_mapper", .k = 1000, .v = std.Some{.x = 5}}: +1
range_mapper_test.RangeMappingKV:
range_mapper_test.RangeMappingKV{.mapper = "test_mapper", .k = 2000, .v = std.Some{.x = 6}}: +1
range_mapper_test.RangeMappingKV:
range_mapper_test.RangeMappingKV{.mapper = "test_mapper", .k = 3000, .v = std.Some{.x = 7}}: +1
range_mapper_test.RangeMappingKV:
range_mapper_test.RangeMappingKV{.mapper = "test_mapper", .k = 4000, .v = std.Some{.x = 8}}: +1
range_mapper_test.RangeMappingKV:
range_mapper_test.RangeMappingKV{.mapper = "test_mapper", .k = 5000, .v = std.Some{.x = 9}}: +1
range_mapper_test.RangeMappingKV:
range_mapper_test.RangeMappingKV{.mapper = "test_mapper", .k = 6000, .v = std.None{}}: +1
